CRM/Birthdays/Mailer.php

```diff
@@ -140,4 +148,29 @@ private function createActivity($target_id, $title, $description): void
             Civi::log()->debug(E::LONG_NAME . ' ' . "Unable to write activity: $exception");
         }
     }
+
+    /**
+     * Returns an array with template data
+     *
+     * e.g.:
+     * [
+     *      'msg_title'
+     *      'msg_subject'
+     *      'msg_html'
+     *      ...
+     * ]
+     *
+     * @throws Exception
+     */
+    private function getMessageTemplate($id): array
+    {
+        try {
+            return \Civi\Api4\MessageTemplate::get()
+                ->addWhere('id', '=', $id)
+                ->execute()
+                ->single();
+        } catch (UnauthorizedException|CRM_Core_Exception $e) {
+            throw new Exception(E::LONG_NAME . " " . "Unable to fetch message template $id. Error: $e");
+        }
```

Comment:
@jensschuppe Is there a example implementation in one of our extensions to fill tokens with content as V3 API is doing it automatically using `MessageTemplate::send` in my previous implementation?

Reply:
API3's `MessageTemplate.send` doesn't return the result, but `CRM_Core_BAO_MessageTemplate::sendTemplate()`, which basically does all the work inside it, does. So maybe it'd be worth not using the API but this method for getting the contents for passing it into the creation of an activity. You might have to adjust parameters, so check what `civicrm_api3_message_template_send()` is doing before.

Rendering the template a second time for creating the activity doesn't seem to make sense to me.